sensor/common/clusterentities/store_endpoints.go

```diff
@@ -0,0 +1,328 @@
+	metrics.UpdateNumberOfEndpoints(len(e.endpointMap), len(e.historicalEndpoints))
+}
+
+// RecordTick records a tick and returns all public IP addresses for which the count should be decremented
+func (e *endpointsStore) RecordTick() bool {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+	removed := false
+	for endpoint, m1 := range e.historicalEndpoints {
+		for deploymentID, m2 := range m1 {
+			for _, status := range m2 {
+				status.recordTick()
+			}
+			e.reverseHistoricalEndpoints[deploymentID][endpoint].recordTick()
+			// Remove all historical entries that expired in this tick.
+			removed = e.removeFromHistoryIfExpired(deploymentID, endpoint) || removed
+		}
+	}
+	return removed
+}
+
+func (e *endpointsStore) Apply(updates map[string]*EntityData, incremental bool) {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+	e.applyNoLock(updates, incremental)
+}
+
+func (e *endpointsStore) applyNoLock(updates map[string]*EntityData, incremental bool) {
+	defer e.updateMetricsNoLock()
+	if !incremental {
+		for deploymentID := range updates {
+			e.purgeNoLock(deploymentID)
+		}
+	}
+	for deploymentID, data := range updates {
+		if data.isDeleteOnly() {
+			// A call to Apply() with empty payload of the updates map (no values) is meant to be a delete operation.
+			continue
+		}
+		e.applySingleNoLock(deploymentID, *data)
+	}
+}
+
+func (e *endpointsStore) purgeNoLock(deploymentID string) set.Set[net.NumericEndpoint] {
+	dec := set.NewSet[net.NumericEndpoint]()
+	// We will be manipulating reverseEndpointMap when calling deleteFromCurrent or moveToHistory,
+	// so let's make a temporary copy
+	endpointsSet := e.reverseEndpointMap[deploymentID]
+	for ep := range endpointsSet {
+		if e.historyEnabled() {
+			// When history expires, the counter will be decremented
+			e.moveToHistory(deploymentID, ep)
+		} else {
+			e.deleteFromCurrent(deploymentID, ep)
+			if ipAddr := ep.IPAndPort.Address; ipAddr.IsPublic() {
+				dec.Add(ep)
+			}
+		}
+	}
+	return dec
+}
+
+func (e *endpointsStore) applySingleNoLock(deploymentID string, data EntityData) {
+	dSet, deploymentFound := e.reverseEndpointMap[deploymentID]
+	if !deploymentFound || dSet == nil {
+		dSet = set.NewSet[net.NumericEndpoint]()
+	}
+
+	for ep, targetInfos := range data.endpoints {
+		dSet.Add(ep)
+		e.reverseEndpointMap[deploymentID] = dSet
+
+		deploymentsOnThisEp, epFound := e.endpointMap[ep]
+		if !epFound {
+			e.endpointMap[ep] = make(map[string]set.Set[EndpointTargetInfo])
+		} else if !deploymentFound {
+			// New deployment, but the endpoint exists - the new deployment takes over the already existing endpoint
+			e.endpointMap[ep][deploymentID] = set.NewSet[EndpointTargetInfo]()
+			// Mark all other deployments having with this endpoint as historical
+			for otherDeploymentID := range deploymentsOnThisEp {
+				// Currently added deployment is already in the map, so do not mark it historical
+				if otherDeploymentID != deploymentID {
+					e.moveToHistory(otherDeploymentID, ep)
+				}
+			}
```

Comment:
Hmm, do I understand correctly the question: we would keep a mapping of a deployment-endpoint to a deployment until a delete command arrives?

In general case, there are many situations that would speak against that, here some examples:
- A deployment can be replaced: nginx IP 1.1.1.1 port 80 is deleted and replaced by apache2 listening on 1.1.1.1 port 80. A lookup request comes to the store "who has 1.1.1.1:80". Now we answer `apache2` because it is not historical, but if there were more than 2 owners of that IP/Port, then the choice would most probably the most recent one from the history. So why to hold more than one in history?
- Most of the calls to the store are done with `incremental=false` which is a different way of saying 'replace`. That always results in a `delete` followed by `apply`. We would need to recognize true-deletes (only delete with no apply) from those replace-deletes we have now. There is a way to call a true-delete, but we use it only in one place.
- The simples reason for not keeping those entries forever is the memory consumption of the stores - it may happen that it would grow for a long time before it is freed - e.g., deployments are moved between the nodes with many updates but never deleted. 

If you asked about something more specific then please let me know and I can refine my answer.

Reply:
> If you asked about something more specific then please let me know and I can refine my answer.

You're answer was perfect 🙂.

> Most of the calls to the store are done with incremental=false which is a different way of saying 'replace.

Yep. I remember noticing that when we were working in re-sync.

While not recommended by k8s, would it be possible for two deployments to share the same endpoint (for example sharing a service)? I think we did a test for two deployments sharing the same IP but different ports. Even if it's possible, I think k8s would have a hard time routing the calls to the correct pod... Maybe it's ok to ignore this edge case.

Reply:
> While not recommended by k8s, would it be possible for two deployments to share the same endpoint (for example sharing a service)? 

I am wondering how would the network traffic routing work for that case 🤔 

Let me not go deeper into this and leave this discussion for posterity in case someone in the future would be handling this edgecase. Remember that we have a special metric for the case that one IP is shared by multiple deployments - that would be a good start for any investigation of this kind.

Reply:
Resolving for now to unlock merging.